tests/src/sanity.rs

```diff
@@ -0,0 +1,326 @@
+        let child_stdout = server.stdout.take().unwrap();
+
+        setup_panic_hook();
+
+        tokio::spawn(async move {
+            loop {
+                let mut error_stream = String::new();
+                // the following is a blocking call, so we need to spawn a task to read the stderr
+                stderr_reader.read_line(&mut error_stream).await.unwrap();
+                // Todo(investigate): it looks like the task reaches the panic when not kept in a
+                // loop i.e. the buffer reads an empty string:  `thread 'main'
+                // panicked at 'Error: '`
+                if !error_stream.is_empty() {
+                    panic!("Error: {}", error_stream);
+                }
+            }
+        });
+
+        // since we are reading from the stdout, we could block at any point in case the server
+        // does not write to its stdout, so we need a timeout here
+        let validation_duration = Duration::from_secs(30);
+        timeout(
+            validation_duration,
+            validate_requests(child_stdout, service_url.as_str(), &mut server),
+        )
+        .await
+        .unwrap();
+
+        // look for the agent job and pod
+        let jobs_api: Api<Job> = Api::namespaced(client.clone(), "default");
+        let jobs = jobs_api.list(&ListParams::default()).await.unwrap();
+        // assuming only one job is running
+        // to make the tests parallel we need to figure a way to get the exact job name when len() >
+        // 1
+        assert_eq!(jobs.items.len(), 1);
+        let job_name = jobs.items[0].metadata.name.clone().unwrap();
+
+        let pods_api: Api<Pod> = Api::namespaced(client.clone(), "default");
+        let pods = pods_api.list(&ListParams::default()).await.unwrap();
+        let pod_name = pods
+            .items
+            .iter()
+            .filter_map(|pod| {
+                if pod.metadata.name.clone().unwrap().contains(&job_name) {
+                    Some(pod.metadata.name.clone().unwrap())
+                } else {
+                    None
+                }
+            })
+            .next()
+            .unwrap();
+        assert!(pod_name.contains(&job_name));
+        sleep(Duration::from_secs(35)).await;
+
+        //verify cleanup
+        let updated_jobs = jobs_api.list(&ListParams::default()).await.unwrap();
+        assert_eq!(updated_jobs.items.len(), 0);
+
+        let updated_pods = pods_api.list(&ListParams::default()).await.unwrap();
+        assert_eq!(updated_pods.items.len(), 1); // only the nginx pod should exist
+    }
+
+    #[tokio::test]
+    // we send a request to a different pod in the cluster (different namespace) and assert
+    // that no operation is performed as specified in the request by the server
+    // as the agent pod is impersonating the pod running in the default namespace
+    async fn test_different_pod_in_cluster() {
+        let client = setup_kube_client().await;
+
```

Comment:
Lets extract all this setup code to an external function instead of repeating it in each test